### ISSUE TITLE
MacOS Big Sur "dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found"

### ISSUE THREAD

**Issue by ChildishhAlbino** [arsenetar/send2trash#51]
![image](https://user-images.githubusercontent.com/36402974/99764313-e5571680-2b50-11eb-8a19-082d5da32110.png)

**Comment by glubsy** [comment]
This is a duplicate of #24.

**Comment by arsenetar** [comment]
Although this is technically a duplicate of #24, the original issue there was due to a different root cause from what I can tell.  This relates to the more recent discussion and this issue probably should be used to track the newer issue as there are two different issues being discussed in #24 now.

**Comment by arsenetar** [comment]
The solutions that I could probably support here are:
1. Compiled module for macOS using probably the NSFileManager as that seems to be the way to interact with the file system without creating an internal finder. Although might have to go the finder route to fix the "put back" issue... from the looks of it.
2. Use PyObjC instead to accomplish the same thing.

@BoboTiG, @HWiese1980,  @acheronfail, or anyone else not sure if there is a preference here or any other opinions as it seems like a change like that is most likely needed to better support these systems.  I would probably set this up similar to the windows implementation in that it would attempt to fall back (and in some cases it would ultimately error if on a system/ python version that has some of the different issues that combine to produce this result).

I might take a look at a quick PyObjC implementation in the next week or so to see how that works out in general as that is probably the fastest for me to get pulled together.

**Comment by HWiese1980** [comment]
Thanks for tagging me here, @arsenetar. I've no explicit preference other than "that it works" and can be installed and used without further ado (except of course `pip install send2trash`). That's basically all I need. Plus love. ;-)

**Comment by andreafrancia** [comment]
@arsenetar:  I've tried using PyObjC for trashing a file but it will not solve the "Put Back" problem.

Here the code:

```Python
import Cocoa
import objc

with open("foo", "a+"):
    pass
path = "foo"
fm = Cocoa.NSFileManager.defaultManager()
fileUrl = Cocoa.NSURL.fileURLWithPath_(path)
result = fm.trashItemAtURL_resultingItemURL_error_(fileUrl, objc.nil, objc.nil)
print(repr(result))
```

To be run with:
```
pip3 install pyobjc-framework-Cocoa
python main.py
```

**Comment by HWiese1980** [comment]
Any good news here so far? @arsenetar

**Comment by arsenetar** [comment]
Got an initial implementation, still need to finish cleaning things up and getting the unit tests updated to test the new implementation as well as the fallback.

**Comment by doogo8** [comment]
Would also like to know when this is going to be fixed :)

**Comment by arsenetar** [comment]
Actually back to working on this, found a bunch of issues in the unit tests across a couple platforms, working on fixing the testing and cleaning it up so that I know there are good tests to validate this against.  Discovered there were more issues there than I had expected.

**Comment by arsenetar** [comment]
Did some testing on macOS again, I wanted to test if the issue was still present in 3.9.2.  I am currently not seeing the issue with python 3.9.2 installed.  I also went back and tested with the system python and currently not having issues with 3.8.2 from the system, this is at macOS 11.2.2.  I'll still finish up the alternative implementation, but not seeing the issue on my machine right now. Also tested with the system's python 2.7.16 and seems to work as well.

**Comment by HWiese1980** [comment]
Awesome news, dude! Keep rocking!

**Comment by johnmcc3** [comment]
I'm still running into this issue on send2trash 1.5.0, macOS 11.2.3, and Python 3.9.2 on an M1 Mac.

**Comment by HWiese1980** [comment]
@johnmcc3 There hasn't been a new release yet, AFAIK.

**Comment by groodt** [comment]
I'm impacted by this transitively through https://github.com/jupyterlab/jupyterlab/issues/9863 There is an unmerged workaround for it https://github.com/jupyter/notebook/pull/6016

Any news on the fix in this repo? Anything I can do to assist?

**Comment by arsenetar** [comment]
@groodt I haven't gotten back to this, need to find time to finish this up, my mac system since installing 3.9.2 does not seem to be throwing this issue anymore so I wont be able to 100% validate the solution on my system.  I should be able to have an update this week.

**Comment by arsenetar** [comment]
I got some code put together for the dual method approach like I have for the windows version its over in the https://github.com/arsenetar/send2trash/tree/feature/%2351 branch at the moment.  I will probably merge this in and package an alpha version in the next day or two. This seems to be working fine on my machine, again mine stopped showing the issue even without the workaround.  If you want to test the workaround you would have to build from that branch and install the `pyobjc-framework-Cocoa` package (with python >= 3.6).

**Comment by arsenetar** [comment]
An alpha labeled package is available now on [pypi](https://pypi.org/project/Send2Trash/1.7.0a0/) can install with `pip install Send2Trash==1.7.0a0` if testing out the pyobjc option need to also install `pyobjc-framework-Cocoa` as well.  I'm going to wait on some feedback on this before marking this as a normal release.

**Comment by HWiese1980** [comment]
Cool! I'll see if I can give it a shot hopefully within the remaining days this week. Can't promise though... If anyone has already, I'll be happy to hear a report ;-)

**Comment by arsenetar** [comment]
@HWiese1980, @groodt, @johnmcc3, @doogo8 have any of you had a chance to verify the alpha version?  I am needing feedback since I cannot reproduce on my system.

**Comment by HWiese1980** [comment]
Not yet, unfortunately. I'll see if I can squeeze it in tomorrow, can't promise though...

**Comment by ITProKyle** [comment]
Can confirm that 1.7.0a0 works on macOS 11.2.3 with python 3.8.5 but only if pyobjc-framework-Cocoa is also installed. Using version 1.5.0 failed with the same setup.

Is the plan to add pyobjc-framework-Cocoa as a dependency (or at least an extra) prior to the final release?

**Comment by ctheune** [comment]
Stumbled over this similar to the other jupyter users, @ITProKyle mentioned it working with 1.7.0a0, but for me it didn't work (3.9, MacOS 11.3.1, didn't matter whether I installed pyobjc-framework-cocoa or not).

**Comment by arsenetar** [comment]
@ITProKyle I am planning on marking pyobjc-framework-Cocoa as an extra for macOS since it is not always required.

@ctheune what was the output when running?

**Comment by ctheune** [comment]
Thanks @arsenetar, here's the traceback:

```ctheune@chazzzzz /tmp/asdf $ bin/jupyter lab
Traceback (most recent call last):
  File "/private/tmp/asdf/bin/jupyter-lab", line 5, in <module>
    from jupyterlab.labapp import main
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab/__init__.py", line 7, in <module>
    from .labapp import LabApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab/labapp.py", line 15, in <module>
    from jupyterlab_server import slugify, WORKSPACE_EXTENSION
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab_server/__init__.py", line 4, in <module>
    from .app import LabServerApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab_server/app.py", line 7, in <module>
    from jupyter_server.extension.application import ExtensionApp, ExtensionAppJinjaMixin
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/extension/application.py", line 21, in <module>
    from jupyter_server.serverapp import ServerApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/serverapp.py", line 75, in <module>
    from jupyter_server.services.contents.filemanager import AsyncFileContentsManager, FileContentsManager
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/services/contents/filemanager.py", line 21, in <module>
    from send2trash import send2trash
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/__init__.py", line 12, in <module>
    from .plat_osx import send2trash
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/plat_osx.py", line 17, in <module>
    GetMacOSStatusCommentString = Foundation.GetMacOSStatusCommentString
  File "/nix/store/hmmqqj3hn6d4lsb3nm07c4pa8qdvvfml-python3-3.9.2/lib/python3.9/ctypes/__init__.py", line 387, in __getattr__
    func = self.__getitem__(name)
  File "/nix/store/hmmqqj3hn6d4lsb3nm07c4pa8qdvvfml-python3-3.9.2/lib/python3.9/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found
```

**Comment by arsenetar** [comment]
@ctheune That trace by the location and content of the issue in `plat_osx.py` seems to indicate you are not using the 1.7.0a0 version of send2trash as the file is now different.

**Comment by ctheune** [comment]
Ouch, sorry, my bad. I just redid the jupyter-install and forgot to do the 1.7.0a0 update. Here it is again:

```
/tmp/asdf $ bin/pip freeze | grep Send
Send2Trash==1.7.0a0
ctheune@chazzzzz /tmp/asdf $ bin/jupyter lab
Traceback (most recent call last):
  File "/private/tmp/asdf/bin/jupyter-lab", line 5, in <module>
    from jupyterlab.labapp import main
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab/__init__.py", line 7, in <module>
    from .labapp import LabApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab/labapp.py", line 15, in <module>
    from jupyterlab_server import slugify, WORKSPACE_EXTENSION
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab_server/__init__.py", line 4, in <module>
    from .app import LabServerApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab_server/app.py", line 7, in <module>
    from jupyter_server.extension.application import ExtensionApp, ExtensionAppJinjaMixin
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/extension/application.py", line 21, in <module>
    from jupyter_server.serverapp import ServerApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/serverapp.py", line 75, in <module>
    from jupyter_server.services.contents.filemanager import AsyncFileContentsManager, FileContentsManager
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/services/contents/filemanager.py", line 21, in <module>
    from send2trash import send2trash
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/__init__.py", line 12, in <module>
    from .plat_osx import send2trash
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/plat_osx.py", line 20, in <module>
    from .plat_osx_ctypes import send2trash  # noqa: F401
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/plat_osx_ctypes.py", line 17, in <module>
    GetMacOSStatusCommentString = Foundation.GetMacOSStatusCommentString
  File "/nix/store/hmmqqj3hn6d4lsb3nm07c4pa8qdvvfml-python3-3.9.2/lib/python3.9/ctypes/__init__.py", line 387, in __getattr__
    func = self.__getitem__(name)
  File "/nix/store/hmmqqj3hn6d4lsb3nm07c4pa8qdvvfml-python3-3.9.2/lib/python3.9/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found
```

**Comment by arsenetar** [comment]
@ctheune That does look like 1.7.0a0 now, Did you also install `pyobjc-framework-Cocoa` looks like it is falling back to the ctypes version.

**Comment by ctheune** [comment]
Sorry, not enough coffee this morning. So this was without `pyobjc-framework-Cocoa`. Here's when I install it:

```
/tmp/asdf $ bin/pip install pyobjc-framework-Cocoa
Collecting pyobjc-framework-Cocoa
  Using cached pyobjc_framework_Cocoa-7.2-cp39-cp39-macosx_10_9_x86_64.whl (273 kB)
Collecting pyobjc-core>=7.2
  Using cached pyobjc_core-7.2-cp39-cp39-macosx_10_9_x86_64.whl (293 kB)
Installing collected packages: pyobjc-core, pyobjc-framework-Cocoa
Successfully installed pyobjc-core-7.2 pyobjc-framework-Cocoa-7.2
WARNING: You are using pip version 20.2.3; however, version 21.1.1 is available.
You should consider upgrading via the '/private/tmp/asdf/bin/python3 -m pip install --upgrade pip' command.
/tmp/asdf $ bin/jupyter lab
Traceback (most recent call last):
  File "/private/tmp/asdf/bin/jupyter-lab", line 5, in <module>
    from jupyterlab.labapp import main
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab/__init__.py", line 7, in <module>
    from .labapp import LabApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab/labapp.py", line 15, in <module>
    from jupyterlab_server import slugify, WORKSPACE_EXTENSION
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab_server/__init__.py", line 4, in <module>
    from .app import LabServerApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyterlab_server/app.py", line 7, in <module>
    from jupyter_server.extension.application import ExtensionApp, ExtensionAppJinjaMixin
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/extension/application.py", line 21, in <module>
    from jupyter_server.serverapp import ServerApp
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/serverapp.py", line 75, in <module>
    from jupyter_server.services.contents.filemanager import AsyncFileContentsManager, FileContentsManager
  File "/private/tmp/asdf/lib/python3.9/site-packages/jupyter_server/services/contents/filemanager.py", line 21, in <module>
    from send2trash import send2trash
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/__init__.py", line 12, in <module>
    from .plat_osx import send2trash
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/plat_osx.py", line 20, in <module>
    from .plat_osx_ctypes import send2trash  # noqa: F401
  File "/private/tmp/asdf/lib/python3.9/site-packages/send2trash/plat_osx_ctypes.py", line 17, in <module>
    GetMacOSStatusCommentString = Foundation.GetMacOSStatusCommentString
  File "/nix/store/hmmqqj3hn6d4lsb3nm07c4pa8qdvvfml-python3-3.9.2/lib/python3.9/ctypes/__init__.py", line 387, in __getattr__
    func = self.__getitem__(name)
  File "/nix/store/hmmqqj3hn6d4lsb3nm07c4pa8qdvvfml-python3-3.9.2/lib/python3.9/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found
```

**Comment by arsenetar** [comment]
@ctheune Looks like the conditional to use pyobjc might be problematic here what do you get if you run the following in the python interpreter in your environment:
~~~ python
from platform import mac_ver
from sys import version_info

print(version_info)
print(mac_ver())
~~~

**Comment by ctheune** [comment]
Here you go:

```
/tmp/asdf $ bin/python
Python 3.9.2 (default, Apr  8 2021, 23:35:33)
[Clang 7.1.0 (tags/RELEASE_710/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from platform import mac_ver
>>> from sys import version_info
>>> print(version_info)
sys.version_info(major=3, minor=9, micro=2, releaselevel='final', serial=0)
>>> print(mac_ver())
('10.16', ('', '', ''), 'x86_64')
>>>
```

**Comment by BoboTiG** [comment]
Ah yes @arsenetar the condition here https://github.com/arsenetar/send2trash/blob/c8bcaea1e8c082d391325834e286bb6b30f25201/send2trash/plat_osx.py#L12 should take into account 10.16.

For compatibility with older binaries, macOS 11.x may report itself as '10.16' rather than '11.x.y'.

**Comment by BoboTiG** [comment]
Sources:
- https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Mac/BuildScript/build-installer.py#L1638-L1640
- https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/test/test_platform.py#L333-L334

Maybe doing a check as done here could be good enough: https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/test/test_posix.py#L1960-L1968

**Comment by arsenetar** [comment]
@BoboTiG Thanks for that information, I did not realize that macOS 11.x did that.  Since this check was mainly just to preserve the existing behavior except when needed I will probably just open up the conditional to allow using the pyobjc interface on any supported version of the OS instead as send2trash itself does not really need to impose any further restrictions.  This would update the conditional to check for 10.9 or newer for the macOS version.  

@ctheune Thanks for trying out the new version and helping work through the issue.  I will try to get an updated pre-release package in the next day or two.

**Comment by ctheune** [comment]
Sure. I'm subscribed here and can test whenever you have something ready. I can test from a branch if you like.

**Comment by arsenetar** [comment]
@ctheune I have updated the pypi package with another pre-release, as with the previous version it is on [pypi](https://pypi.org/project/Send2Trash/1.7.0a1/) and can be installed with `pip install Send2Trash==1.7.0a1`.

**Comment by ctheune** [comment]
Aaaand we have liftoff. I've updated my existing instance and now it runs fine. Just to make sure: `pyobjc-framework-Cocoa` is still required and uninstallting it will fall back to the ctypes version which still fails but - but that's intentional, right?

**Comment by arsenetar** [comment]
Yes `pyobjc-framework-Cocoa` is still required on systems running into the issue, I let it fall back to the ctypes version just in case it does work on the platform (as is the case on mine for some reason...).  I need to update some of the documentation to make this clearer.

**Comment by arsenetar** [comment]
I'll be moving this to a stable release in the next week since it seems this solution is working.

**Comment by bergkvist** [comment]
Yeah, doing the following fixed the problem for me:
```
pip install pyobjc-framework-Cocoa==7.3 send2trash==1.7.0a1
```

**Comment by arsenetar** [comment]
Finally published as 1.7.1 on pypi.
